File: Python/utilities/get_gpfr.py
import numpy as np
from scipy.stats import multivariate_normal


def get_gpfr(fr, xx):
    binwidth = xx[4] - xx[3]
    tau = 100 * binwidth
    sigm = 1

    numtr = fr.shape[0]
    gpmean = np.nanmean(fr) + np.zeros(len(xx))

    kernel_fun = np.zeros((len(xx), len(xx)))
    for tii in range(len(xx)):
        for tjj in range(len(xx)):
            ti = xx[tii]
            tj = xx[tjj]
            kernel_fun[tii, tjj] = sigm**2 * np.exp(-(ti - tj)**2 / (2 * tau**2))

    gpfr = multivariate_normal.rvs(mean=gpmean, cov=kernel_fun, size=numtr)
    return gpfr, kernel_fun

File: MATLAB/utilities/get_gpfr.m
function [gpfr,kernel_fun]=get_gpfr(fr,xx )

%given a matrix of firing rate (num trials x time) and time samples, this function creates a
%gaussian process surrogate data of same size as fr with mean at global
%mean of fr, tau of 100ms and sigma of 1. 
%tutorial for GP: https://peterroelants.github.io/posts/gaussian-process-tutorial/



binwidth=xx(5)-xx(4);
%params for GP surrogate data:
tau=100*binwidth; %100ms in order to match the gaussing smoothing applied to real psth
sigm=1;

numtr=size(fr,1);
gpmean = nanmean(fr(:))+zeros(size(xx));
for tii=1:length(xx)
    for tjj=1:length(xx)
        ti=xx(tii);
        tj=xx(tjj);
        kernel_fun(tii,tjj) = sigm^2*exp(-(ti-tj)^2/(2*tau^2));
    end
end

gpfr = mvnrnd(gpmean,kernel_fun(:,:),numtr);

% % visualize your surrogate data:
% subplot(2,2,1);plot(xx,gpfr');title 'GP time course for each trial'
% subplot(2,2,2);imagesc(xx,1:numtr,gpfr); xlabel time; ylabel trials; title GP
% subplot(2,2,3);imagesc(kernel_fun); title 'covariance matrix'
end






